app/questionnaire/placeholder_parser.py

```diff
@@ -40,50 +38,85 @@ def __init__(
         language: str,
         answer_store: AnswerStore,
         list_store: ListStore,
-        metadata: Optional[MetadataProxy],
+        metadata: MetadataProxy | None,
         response_metadata: Mapping,
         schema: QuestionnaireSchema,
         renderer: "PlaceholderRenderer",
-        list_item_id: Optional[str] = None,
-        location: Union[Location, RelationshipLocation, None] = None,
-        placeholder_preview_mode: Optional[bool] = False,
+        progress_store: ProgressStore,
+        list_item_id: str | None = None,
+        location: Location | RelationshipLocation | None = None,
+        placeholder_preview_mode: bool | None = False,
     ):
+        self._transformer = PlaceholderTransforms(language, schema, renderer)
+        self._placeholder_map: MutableMapping[
+            str, Union[ValueSourceEscapedTypes, ValueSourceTypes, None]
+        ] = {}
         self._answer_store = answer_store
         self._list_store = list_store
         self._metadata = metadata
         self._response_metadata = response_metadata
-        self._schema = schema
         self._list_item_id = list_item_id
+        self._schema = schema
         self._location = location
-        self._transformer = PlaceholderTransforms(language, schema, renderer)
-        self._placeholder_map: MutableMapping[
-            str, Union[ValueSourceEscapedTypes, ValueSourceTypes, None]
-        ] = {}
+        self._progress_store = progress_store
         self._placeholder_preview_mode = placeholder_preview_mode
 
-        self._value_source_resolver = ValueSourceResolver(
+        self._path_finder = pf.PathFinder(
+            schema=self._schema,
             answer_store=self._answer_store,
             list_store=self._list_store,
+            progress_store=self._progress_store,
             metadata=self._metadata,
-            schema=self._schema,
-            location=self._location,
-            list_item_id=self._list_item_id,
-            escape_answer_values=True,
             response_metadata=self._response_metadata,
-            use_default_answer=True,
         )
 
+        self._value_source_resolver = self._get_value_source_resolver()
+        self._routing_paths: dict = {}
```

Comment:
Minor, for clarity, I would rename to `self._routing_path_block_ids_by_section_key`

Reply:
Updated 👍